app/src/main/java/com/eveningoutpost/dexdrip/Services/DexCollectionService.java

```diff
@@ -224,76 +280,189 @@ public void onServicesDiscovered(BluetoothGatt gatt, int status) {
                 BluetoothGattService gattService = mBluetoothGatt.getService(xDripDataService);
                 if (gattService != null) {
                     BluetoothGattCharacteristic gattCharacteristic = gattService.getCharacteristic(xDripDataCharacteristic);
-                    if (gattCharacteristic != null ) {
+                    if (gattCharacteristic != null) {
                         final int charaProp = gattCharacteristic.getProperties();
-
                         if ((charaProp | BluetoothGattCharacteristic.PROPERTY_NOTIFY) > 0) {
-                            mBluetoothGatt.setCharacteristicNotification(gattCharacteristic, true);
+                            if ((charaProp | BluetoothGattCharacteristic.PROPERTY_NOTIFY) > 0) {
+                                mBluetoothGatt.setCharacteristicNotification(gattCharacteristic, true);
+                            } else {
+                                Log.e(TAG, "onServicesDiscovered: characteristic " + gattCharacteristic.getUuid() + " doesn't have notify properties");
+                            }
                         } else {
-                            Log.e(TAG, "characteristic " + gattCharacteristic.getUuid() + " doesn't have notify properties");
+                            Log.e(TAG, "onServicesDiscovered: characteristic " + xDripDataCharacteristic + " not found");
                         }
                     } else {
-                        Log.e(TAG, "characteristic " + xDripDataCharacteristic + " not found");
+                        Log.e(TAG, "onServicesDiscovered: service " + xDripDataService + " not found");
+                        //Log.v(TAG, "onServicesDiscovered: Already have that service");
                     }
-                } else {
-                    Log.e(TAG, "service " + xDripDataCharacteristic + " not found");
+                    Log.w(TAG, "onServicesDiscovered received success: " + status);
                 }
+            } else {
+                Log.w(TAG, "onServicesDiscovered received: " + status);
             }
         }
 
         @Override
-        public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
-            final byte[] data = characteristic.getValue();
-            if (data != null && data.length > 0) { setSerialDataToTransmitterRawData(data, data.length); }
+        public void onCharacteristicRead(BluetoothGatt gatt,
```

Comment:
Happy to entertain a better way, but it doesn't cause duplicate data. Been running well. 

 
John Stevens
M - +61 418526349

> On 20 May 2015, at 00:01, Stephen Black <notifications@github.com> wrote:
> 
> In app/src/main/java/com/eveningoutpost/dexdrip/Services/DexCollectionService.java:
> 
> >              }
> >          }
> >  
> >          @Override
> > -        public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
> > -            final byte[] data = characteristic.getValue();
> > -            if (data != null && data.length > 0) { setSerialDataToTransmitterRawData(data, data.length); }
> > +        public void onCharacteristicRead(BluetoothGatt gatt,
> Wont this always cause it to try and duplicate data? why are we using onCharacteristicRead when the only time we are reading it is when we are already trying to pull it out of the queue, this will have it attempt to create duplicate data, I addressed this on the other branch as well.
> 
> —
> Reply to this email directly or view it on GitHub.
>